### DISCUSSION TITLE
Is `typing.Protocol` the only way to concatenate the keyword-only arguments in `ParamSpec`?

### DISCUSSION THREAD

**Discussion by McDic** [python/typing#1905] [Q&A] [answered]
Hello, I wrote a decorator code adds custom keyword-only argument on the function using `typing.Protocol` and generics.

```python
import typing

P = typing.ParamSpec("P")
Ret = typing.TypeVar("Ret", covariant=True)  # Mypy complains this should be covariant


class CustomFunctionProtocol(typing.Protocol[P, Ret]):
    def __call__(
        self,
        *args: P.args,
        special_kwarg: str,
        **kwargs: P.kwargs,
    ) -> Ret: ...


def decorator(f: typing.Callable[P, Ret]) -> CustomFunctionProtocol[P, Ret]:

    def inner_func(*args: P.args, special_kwarg: str, **kwargs: P.kwargs) -> Ret:
        print("Special kwarg given: ", special_kwarg)
        return f(*args, **kwargs)

    return inner_func


@decorator
def example_function(a: int, b: int, c: int) -> int:
    return a + b + c


typing.reveal_type(example_function)
```

In this example code, `decorator` adds a keyword-only argument `special_kwarg` to a decorated function, type hint provided by `CustomFunctionProtocol`.

When I run `mypy` for this Python code, there is no error. I see `reveal_type(example_function)` is also working well.

```
Type of "example_function" is "CustomFunctionProtocol[(a: int, b: int, c: int), int]" [Pylance]

Revealed type is "typing_test_20241224.CustomFunctionProtocol[[a: builtins.int, b: builtins.int, c: builtins.int], builtins.int]" [Mypy]
```

However, running `pyright` will report an error regarding `special_kwarg`;

```
typing_test_20241224.py:11:9 - error: Keyword parameter "special_kwarg" cannot appear in signature after ParamSpec args parameter (reportGeneralTypeIssues)
typing_test_20241224.py:18:35 - error: Keyword parameter "special_kwarg" cannot appear in signature after ParamSpec args parameter (reportGeneralTypeIssues)
typing_test_20241224.py:30:20 - information: Type of "example_function" is "CustomFunctionProtocol[(a: int, b: int, c: int), int]"
```

When I try to use that function on code, VSCode intellisense does not reflect the `ParamSpec` variable properly, it shows `example_function(*args, special_kwarg: str, **kwargs)` instead. I installed Mypy type checker, Pylance and Python extensions.

![screenshot2](https://github.com/user-attachments/assets/ce6657a3-fe8b-4032-b92d-4d77e7162916)

Perhaps showing correct intellisense for generic `Protocol`'s `__call__` is not implemented yet. I do not know where to ask or complain this, neither I know how to make this intellisense do the correct behavior, so I am seeking if there is any other way to give a type hint for "kwargs" version of `typing.Concatenate`.

**Comment by Daverball** [comment, marked as answer]
There's no way to do this currently, even the `Protocol` way doesn't work and pyright is correct to complain. You are only allowed prepend positional only parameters using `typing.Concatenate` and `*args: P.args` must always be directly followed by `**kwargs: P.kwargs` in a signature.

You can perform slightly more complex parameter list extensions using `TypeVarTuple`, but this is once again only useful for positional-only arguments.

This was a [deliberate design choice of PEP-612](https://peps.python.org/pep-0612/#concatenating-keyword-parameters) to keep the scope of the change smaller and easier to understand and implement. Even with the simplification it proved to be one of the most challenging changes for type checkers to implement. It took quite a while for support in mypy to land and in pytype there's still no full `ParamSpec` support and virtually no `TypeVarTuple` support at all to this day.

`ParamSpec` is mature enough at this point, that someone could take on the work of specifying some of the more advanced use-cases like upper bounds and complex argument augmentations. But someone has to actually do the required work of figuring out the detailed semantics and do it well enough for it to be accepted. So far, nobody has seemed interested enough to take it upon themselves.
